consensus/src/consensus_observer/observer/subscription_manager.rs

```diff
@@ -347,38 +169,89 @@ impl SubscriptionManager {
-    /// Note: if `unhealthy_subscription_peers` are provided, they will be excluded
-    /// from the selection process. Likewise, all peers currently subscribed to us
-    /// will be excluded from the selection process.
-    fn sort_peers_for_subscriptions(
+    /// Spawns a new subscription creation task to create
+    /// the specified number of new subscriptions.
+    async fn spawn_subscription_creation_task(
         &mut self,
-        mut connected_peers_and_metadata: HashMap<PeerNetworkId, PeerMetadata>,
-        unhealthy_subscription_peers: Vec<PeerNetworkId>,
-    ) -> Option<Vec<PeerNetworkId>> {
-        // Remove any peers we're already subscribed to
-        for active_subscription_peer in self.get_active_subscription_peers() {
-            let _ = connected_peers_and_metadata.remove(&active_subscription_peer);
+        num_subscriptions_to_create: usize,
+        active_subscription_peers: Vec<PeerNetworkId>,
+        terminated_subscriptions: Vec<(PeerNetworkId, Error)>,
+        connected_peers_and_metadata: HashMap<PeerNetworkId, PeerMetadata>,
+    ) {
+        // If there are no new subscriptions to create, return early
+        if num_subscriptions_to_create == 0 {
+            return;
         }
 
-        // Remove any unhealthy subscription peers
-        for unhealthy_peer in unhealthy_subscription_peers {
-            let _ = connected_peers_and_metadata.remove(&unhealthy_peer);
+        // If there is an active subscription creation task, return early
+        if let Some(subscription_creation_task) = &*self.active_subscription_creation_task.lock() {
+            if !subscription_creation_task.is_finished() {
+                return; // The task is still running
+            }
         }
 
-        // Remove any peers that are currently subscribed to us
-        if let Some(consensus_publisher) = &self.consensus_publisher {
-            for peer_network_id in consensus_publisher.get_active_subscribers() {
-                let _ = connected_peers_and_metadata.remove(&peer_network_id);
+        // Clone the shared state for the task
+        let active_observer_subscriptions = self.active_observer_subscriptions.clone();
+        let consensus_observer_config = self.consensus_observer_config;
+        let consensus_observer_client = self.consensus_observer_client.clone();
+        let consensus_publisher = self.consensus_publisher.clone();
+        let db_reader = self.db_reader.clone();
+        let time_service = self.time_service.clone();
+
+        // Spawn a new subscription creation task
+        let subscription_creation_task = tokio::spawn(async move {
+            // Identify the terminated subscription peers
+            let terminated_subscription_peers = terminated_subscriptions
+                .iter()
+                .map(|(peer, _)| *peer)
+                .collect();
+
+            // Create the new subscriptions
+            let new_subscriptions = subscription_utils::create_new_subscriptions(
```

Comment:
if there's two subscriptions to create, this is still sequential inside?

Reply:
Yes 😃 We still want to keep the ordering when creating subscriptions, and we need to be able to handle potential failures, so it's simplest to keep them sequential 😄 We can always parallelize this if we want, but for only 2 subscriptions, it's probably okay.